File: algorithms/solutions/week24/taurandat/coin-change.cpp
/*
 *
 * Link: https://leetcode.com/problems/coin-change
 *
 * 182 / 182 test cases passed.
 * Status: Accepted
 * Runtime: 34 ms
 * Your runtime beats 46.56 % of cpp submissions.
 *
 */

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int Max = amount + 1;
        vector<int> dp(amount + 1, Max);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.size(); j++) {
                if (coins[j] <= i) {
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1);
                }
            }
        }

        return dp[amount] > amount ? -1 : dp[amount];
    }
};

File: algorithms/solutions/week24/duythinht/coin_change.go
package duythinht

//https://leetcode.com/problems/sum-of-square-numbers/description/

func coinChange(coins []int, amount int) int {

	exchangeTable := make([]int, amount+1)

	for value := 1; value < len(exchangeTable); value++ {
		exchangeTable[value] = -1
		for _, coin := range coins {
			if coin > value {
				continue
			}
			count := 1 + exchangeTable[value-coin]
			if count > 0 && (count < exchangeTable[value] || exchangeTable[value] < 0) {
				exchangeTable[value] = count
			}
		}
	}
	return exchangeTable[amount]
}
